### PR TITLE
4050: Interactive components

### PR DESCRIPTION
**Pull request by kat1906**

### CODE REVIEW

File: app/views/includes/layout.njk [hunk 1]
```diff
@@ -31,15 +31,15 @@
     <meta name="msapplication-square310x310logo" content="/nhsuk-frontend/assets/favicons/largetile-310x310.png">
   </head>
 
-  <body>
+  <body style="{height:100% !important;}">
```
Comment (BenBJSS) [inline app/views/includes/layout.njk]: Should we create a CSS class for this?

**Comment by kat1906** [comment]
> You'll need to update the package.json version

Doh!

File: app/views/includes/layout.njk (hunk 1 above)
Comment (kat1906) [reply app/views/includes/layout.njk]: yes - forgot about this :D

**Comment by alibombaye** [comment]
> > You'll need to update the package.json version
> 
> Doh!

on this. Do you think this is sufficient enough change to warrant changing it to 1.1.0?

File: app/helpers/defaultParams.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  },
+  'view-section-table-row': {
+    componentName: 'viewSectionTableRow',
+    params: {
+      questionId: 'some-question-id',
+      questionTitle: 'Some question title',
+      layout: 'vertical',
+      innerComponent: '&ltdiv&gt;Some inner component&lt/div&gt;',
+    },
+  },
```
Comment (alibombaye) [inline app/helpers/defaultParams.js]: mmm I would like these default params for each of these to live within their component folder if possible please 😏

File: app/routes.js [hunk 3]
```diff
@@ -1,14 +1,47 @@
...
+  res.render('template');
+});
+
+router.post('/component/:component', async (req, res) => {
+  const { component } = req.params;
+  const params = Object.entries(req.body).reduce((acc, [param, value]) => {
+    if (!value) acc[param] = defaultParams[component].params[param];
+    else acc[param] = value;
+    return acc;
+  }, {});
```
Comment (alibombaye) [inline app/routes.js]: Could we have a helper function that describes what's going on here?

File: app/routes.js [hunk 4]
```diff
@@ -1,14 +1,47 @@
...
+
+  if (component === 'view-data-bulletlist') {
+    params.data = params.data.split(',');
+  }
```
Comment (alibombaye) [inline app/routes.js]: could we move this withing the helper thing above? or the generateTemplate function?

File: app/helpers/defaultParams.js (hunk 2 above)
Comment (kat1906) [reply app/helpers/defaultParams.js]: ok

File: .gitignore [hunk 5]
```diff
@@ -63,3 +63,6 @@ jspm_packages/
 # System Files
 .DS_Store
 Thumbs.db
+
+# generated templates
+app/templates/
```
Comment (alibombaye) [inline .gitignore]: 👍

File: app/components/view-data-title/settings.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+  "componentName": "viewDataTitle",
+  "params": {
+    "title": "Title",
+    "classes": "another-class"
+  }
+}
```
Comment (alibombaye) [inline app/components/view-data-title/settings.json]: complaining about no newline. You might as well add it

File: app/helpers/templateGenerator.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      ? (`${opts.stringStart}${key}: ${JSON.stringify(modifiedValue)}`)
+      : (`${opts.stringStart0}${key}: ${JSON.stringify(modifiedValue)}`);
+    acc[0] += keyValueString;
+    return acc;
+  }, opts.accArray).join('');
+};
+
+const getSettings = (component) => {
+  const settingsString = fs.readFileSync(`app/components/${component}/settings.json`, 'utf-8');
+  return JSON.parse(settingsString);
```
Comment (alibombaye) [inline app/helpers/templateGenerator.js]: Nice 😄

**Comment by kat1906** [comment]
> Awesome! can't wait to see the interactive sections next

😨

**Pull request merged**